src/index.js

```diff
@@ -46,28 +46,46 @@ function hasParams(required, actual) {
 }
 
 /**
- * Create providers along index definitions and return an array of objects
- * consisting of an index definition and its providers
+ * Create handlers along index definitions and return an array of objects
+ * consisting of an index definition and its handler.
  *
- * @param {Array} indices array of index definitions
- * @returns array of index definitions and their providers
+ * @param {Array} configs array of index definitions
+ * @param {object} params parameters available
+ * @param {object} log logger
+ *
+ * @returns array of index definitions and their handlers
  */
-function createProviders(indices, params, log) {
-  return indices
-    .map((config) => {
-      let provider;
-      const candidate = providers
-        .filter((c) => hasParams(c.required, params))
-        .find((c) => c.match(config.target));
-      if (candidate) {
-        try {
-          provider = candidate.create(params, config, log);
-        } catch (e) {
-          log.error(`Unable to create provider for ${config.name}`, e);
-        }
+function createHandlers(configs, params, log) {
+  const configMap = configs
+    .map((config) => providers
+      .filter((provider) => hasParams(provider.required, params))
+      .find((provider) => provider.match(config.target)))
+    .reduce((map, provider, i) => {
+      const name = provider ? provider.name : '(none)';
+      if (!map[name]) {
+        // eslint-disable-next-line no-param-reassign
+        map[name] = { provider, configs: [], indices: [] };
+      }
+      map[name].indices.push(i);
+      map[name].configs.push(configs[i]);
+      return map;
+    }, {});
+
+  const result = new Array(configs.length);
+  Object.values(configMap).forEach(({ provider, configs: providerConfigs, indices }) => {
+    let handlers = new Array(indices.length);
+    try {
+      if (provider) {
+        handlers = provider.create(params, providerConfigs, log);
       }
-      return { config, provider };
+    } catch (e) {
+      log.error(`Unable to create handler in ${provider.name}`, e);
+    }
+    indices.forEach((index, i) => {
+      result[index] = { config: configs[index], handler: handlers[i] };
     });
+  });
+  return result;
```

Comment:
For composite providers (such as Excel), it is necessary to create a handler per subtree, but still have a parent managing those handlers, so it can collect changes made to its subtrees and send just *one* message to the queue.